### PR TITLE
[suggestion] add interrupt() to stop event

### PR DESCRIPTION
**Pull request by mattn**
In current implementation, we can't stop writing file in BufWritePre without send throw. For example, if we can to avoid typo of filename to write, no way to stop writing.

```
au BufWritePre * if expand('<amatch>:t') == ']' | throw "oops!!!!!!!!!!" | endif
```

I know that the policy of vim is that use BufWriteCmd to checks for mistakes in filenames and quit with error messages (without writing the file). But I do not think this is seamless. The implementation we should have is to be able to cancel the event. So I want to suggest this PR that add cancelevent(). This function can cancel XxxxxxPre event. This is similar to typing CTRL-C. And this function do nothing if it is callied in non-event function.

How do you think?

### CODE REVIEW

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> In current implementation, we can't stop writing file in BufWritePre
> without send throw. For example, if we can to avoid typo of filename
> to write, no way to stop writing.
> 
> ```
> au BufWritePre * if expand('<amatch>:t') == ']' | throw "oops!!!!!!!!!!" | endif
> ```
> 
> I know that the policy of vim is that use BufWriteCmd to checks for
> mistakes in filenames and quit with error messages (without writing
> the file). But I do not think this is seamless. The implementation we
> should have is to be able to cancel the event. So I want to suggest
> this PR that add cancelevent(). This function can cancel XxxxxxPre
> event. This is similar to typing CTRL-C. And this function do nothing
> if it is callied in non-event function.
> 
> How do you think?

I like the idea, but cancelevent() is too specific.

How about a function that works like hitting CTRL-C?  It would set
"got_int", which means most operations bail out.  We can call it
interrupt().

-- 
If you're sending someone Styrofoam, what do you pack it in?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
implemented f_interrupt() and added test & doc.

File: src/testdir/test_interrupt.vim [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+endfunction
+
+func Test_interrupt()
+  new Xfile
+  let n = 0
+  au BufWritePre Xfile call s:bufwritepre()
+  au BufWritePost Xfile call s:bufwritepost()
+  w!
+  call assert_equal(1, s:bufwritepre_called)
+  call assert_equal(1, s:bufwritepost_called)
```
Comment (mattn) [inline src/testdir/test_interrupt.vim]: Must be 0

Comment (mattn) [reply src/testdir/test_interrupt.vim]: Ah, tests stop with Vim:Interrupt. Hmm.

**Pull request closed without merging**